### PR TITLE
Add new scheduler of load in fe

### PR DESCRIPTION
**Pull request by EmmyMiao87**
1. New scheduler only support the broker load now.
2. The stage of load consist of PENDING -> LOADING -> FINISHED
3. The LoadScheduler will divide job into a pending task. There are preparations that need to be done on pending task.
4. OnPendingTaskFinished will be invoked after pending task. It is used to submit the loading task which is created based on attachment of pending task.
5. OnLoadingTaskFinished will be invoked after loding task. It is used to record the commit info and commit txn when all of task has been finished.
.

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadManager.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+/*
+This class contains all of broker and mini load job(v2).
+ */
+public class LoadManager {
+    private Map<Long, LoadJob> idToLoadJob = Maps.newConcurrentMap();
+    private MasterTaskExecutor pendingTaskExecutor =
+            new MasterTaskExecutor(Config.load_pending_thread_num_high_priority);
+    private MasterTaskExecutor loadingTaskExecutor = new MasterTaskExecutor(10);
+
+    private ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/LoadManager.java]: What is this lock for?

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadingTaskAttachment.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+package org.apache.doris.load.loadv2;
+
+import org.apache.doris.transaction.TabletCommitInfo;
+
+import java.util.List;
+import java.util.Map;
+
+public class BrokerLoadingTaskAttachment implements LoadLoadingTaskAttachment {
+
+    private Map<String, Long> fileMap;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadingTaskAttachment.java]: Plz add comment to explain the key and value of the map

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadScheduler.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+If there is a pending task belong to pending job, job will not be rescheduled.
+ */
+public class LoadScheduler extends Daemon {
+
+    private static final Logger LOG = LogManager.getLogger(LoadScheduler.class);
+
+    private LoadManager loadManager;
+
+    public LoadScheduler(LoadManager loadManager) {
+        super();
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/LoadScheduler.java]: Better adding a thread name for debugging easily

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadTaskCallback.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.doris.load.loadv2;
+
+public interface LoadTaskCallback {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/LoadTaskCallback.java]: wen can make this callback more general.

I think that define only `onTaskFinished` and `onTaskFailed` is enough and give job more options

**Review by morningman — CHANGES_REQUESTED** [review]

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadManager.java (hunk 1 above)
Comment (EmmyMiao87) [reply fe/src/main/java/org/apache/doris/load/loadv2/LoadManager.java]: To protect the atomicity of (idToLoadJob, dbIdToLabelToLoadJobs)

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadingTaskAttachment.java (hunk 2 above)
Comment (EmmyMiao87) [reply fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadingTaskAttachment.java]: I will remove it later.

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadLoadingTask.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        this.table = table;
+        this.brokerDesc = brokerDesc;
+        this.fileGroups = fileGroups;
+        this.jobDeadlineMs = jobDeadlineMs;
+        this.execMemLimit = execMemLimit;
+        this.txnId = txnId;
+        this.callback = callback;
+    }
+
+    public void init(List<List<TBrokerFileStatus>> fileStatusList, int fileNum) throws UserException {
```
[the remark this answers is not in the document]
Comment (EmmyMiao87) [reply fe/src/main/java/org/apache/doris/load/loadv2/LoadLoadingTask.java]: The db lock has been held in the BrokerLoadJob.

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+                // Generate loading task and init the plan of task
+                LoadLoadingTask task = new LoadLoadingTask(db, table, brokerDesc,
+                                                           entry.getValue(), getDeadlineMs(), execMemLimit,
+                                                           transactionId, this);
+                task.init(attachment.getFileStatusByTable(tableId),
+                          attachment.getFileNumByTable(tableId));
+                // Add tasks into list and pool
+                tasks.add(task);
+                Catalog.getCurrentCatalog().getLoadManager().submitTask(task);
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: Can you pass loadManager when this object is created?
we should avoid using getInstance() in our code

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        Catalog.getCurrentCatalog().getLoadManager().submitTask(task);
+    }
+
+    @Override
+    public void onTaskFinished(TaskAttachment attachment) {
+        if (attachment instanceof BrokerPendingTaskAttachment) {
+            onPendingTaskFinished((BrokerPendingTaskAttachment) attachment);
+        } else if (attachment instanceof BrokerLoadingTaskAttachment) {
+            onLoadingTaskFinished((BrokerLoadingTaskAttachment) attachment);
+        }
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: Do you have some `state` or something to indicate whether or not this call is valid in your state machine?

For example this function can be called concurrently? or called multiple times for same task?

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        // TODO(ml): check if task has been cancelled
+        boolean commitTxn = false;
+        writeLock();
+        try {
+            // update loading status
+            updateLoadingStatus(attachment);
+
+            // begin commit txn when all of loading tasks have been finished
+            if (tasks.size() == tasks.stream()
+                    .filter(entity -> entity.isFinished()).count()) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: why not use a set to record which task has finished?

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                    commitTxn = true;
+                }
+            }
+        } finally {
+            writeUnlock();
+        }
+
+        Database db = null;
+        try {
+            getDb();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: db = getDb()?

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java (hunk 7 above)
Comment (EmmyMiao87) [reply fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: Situation 1 (BrokerPendingTask ):There is only one pending task in BrokerLoad. So this function will not be called concurrently when pending task.
Situation 2 (LoadingTask): I add the writelock before onLoadingTaskFinished. So the loading task will be serial executed.

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+                OlapTable table = (OlapTable) db.getTable(tableId);
+                if (table == null) {
+                    LOG.warn(new LogBuilder(LogKey.LOAD_JOB, id)
+                                     .add("database_id", dbId)
+                                     .add("table_id", tableId)
+                                     .add("error_msg", "Failed to divide job into loading task when table not found")
+                                     .build());
+                    cancelJobWithoutCheck(FailMsg.CancelType.ETL_RUN_FAIL,
+                                          "Unknown table(" + tableId + ") in database(" + db.getFullName() + ")");
+                    return;
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]: throw Exception?

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+                             .add("error_msg", "Failed to divide job into loading task.")
+                             .build(), e);
+            cancelJobWithoutCheck(FailMsg.CancelType.ETL_RUN_FAIL, e.getMessage());
+            return;
+        }
+
+        loadStartTimestamp = System.currentTimeMillis();
+    }
+
+    private void allocateLoadingTask(Database db, BrokerPendingTaskAttachment attachment) throws UserException {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java]:
```suggestion
    private void createLoadingTask(Database db, BrokerPendingTaskAttachment attachment) throws UserException {
```

File: fe/src/main/java/org/apache/doris/load/loadv2/LoadJobScheduler.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private void process() throws InterruptedException {
+        while (!needScheduleJobs.isEmpty()) {
+            // take one load job from queue
+            LoadJob loadJob = needScheduleJobs.take();
+
+            // schedule job
+            try {
+                loadJob.scheduleJob();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/LoadJobScheduler.java]: scheduler call `scheduleJob`? a little strange

File: fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadPendingTask.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            List<List<TBrokerFileStatus>> fileStatusList = Lists.newArrayList();
+            List<BrokerFileGroup> fileGroups = entry.getValue();
+            for (BrokerFileGroup fileGroup : fileGroups) {
+                List<TBrokerFileStatus> fileStatuses = Lists.newArrayList();
+                for (String path : fileGroup.getFilePathes()) {
+                    BrokerUtil.parseBrokerFile(path, brokerDesc, fileStatuses);
+                }
+                fileStatusList.add(fileStatuses);
+                for (TBrokerFileStatus fstatus : fileStatuses) {
+                    if (LOG.isDebugEnabled()) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadPendingTask.java]: put this judge out of `for` block?

**Review by imay — APPROVED** [review]

**Review by morningman — APPROVED** [review]

**Pull request merged**